sdks/java/core/src/main/java/org/apache/beam/sdk/io/Write.java

```diff
@@ -151,7 +145,31 @@ public int getNumShards() {
      * runner-controlled sharding.
      */
     public Bound<T> withNumShards(int numShards) {
-      return new Bound<>(sink, Math.max(numShards, 0));
+      if (numShards <= 0) {
+        return new Bound<>(sink, null);
+      } else {
+        return new Bound<>(sink, new ConstantShards<T>(numShards));
+      }
+    }
+
+    /**
+     * Returns a new {@link Write.Bound} that will write to the current {@link Sink} using the
+     * specified {@link PTransform} to compute the number of shards.
+     *
+     * <p>This option should be used sparingly as it can hurt performance. See {@link Write} for
+     * more information.
+     */
+    public Bound<T> withSharding(PTransform<PCollection<T>, PCollectionView<Integer>> sharding) {
+      checkNotNull(sharding, "Cannot provide null sharding. Use withoutSharding() instead");
+      return new Bound<>(sink, sharding);
+    }
+
+    /**
+     * Returns a new {@link Write.Bound} that will write to the current {@link Sink} with
+     * runner-determined sharding.
+     */
+    public Bound<T> withoutSharding() {
```

Comment:
`withRunnerDeterminedSharding` is more precise (as DynamicSharding can be user-determined)